## <i>Oculina arbuscula</i> gene expression study between aposymbiotic and symbiotic branches
This repo contains the data and scripts associated with the manuscript: 
Rivera, H.E. & Davies, S.W. (in prep) Symbiosis maintenance in the facultative coral <i>Oculina arbuscula</i> relies on nitrogen cycling, cell cycle modulation, and immunity

### Organization
The repository is separated into scripts, data, figures, and tables folders. 

The scripts folder contains all scripts that were used to: 
1. Map RNA-seq reads to the coral and symbiont transcriptomes (read_mapping.sh). For transcritome assembly scripts see: https://github.com/hrivera28/Oculina_arbuscula_transcriptome.
2. Compare mapping efficiencies of symbiotic and aposymbiotic samples to coral and symbiont transcriptome and generate figure (fig_S2.R)
3. Call snps on RNA seq data to test for clones (oculina_snps.sh and oculina_snps.R) 
3. Process counts of mapped reads to determine differential expression between symbiotic states and conduct other analyses and generate figures (deg_analyses.Rmd). There is a well commented R markdown html file that contains all the output embedded. 
4. Generate map of coral habitat range (fig_1.R) 

All the input data that would be needed to run these scripts is included in the data folder. Intermediate files that can be generated by running the scripts with the included input data are not included. 

Final figures and tables that are in the manuscript and supplementary data are included in the respective figures and tables folders for easy reference. 
